### PR TITLE
docs(indexing): add proposal for sourceHash property

### PR DESCRIPTION
**Pull request by dominique-pfister**
New proposal to remove extension automatically and add a `sourceHash` provider property.

Just wondering: couldn't we just call that thing `id`, and request its uniqueness, whatever its underlying semantics?

### CODE REVIEW

**Comment by tripodsan** [comment]
> Just wondering: couldn't we just call that thing `id`, and request its uniqueness, whatever its underlying semantics?

good idea. but maybe `uid` in order to underline that it should not be human readable (the idea of the hash was, that the source (e.g. document name) is not visible as-is in the HTML).

File: docs/observation/payload.md [hunk 1]
```diff
@@ -1,34 +1,52 @@
 # Format of observation payload
 
-Notifications received by a source repository, such as GitHub, OneDrive, Google Drive, etc. have a provider specific format, e.g. OneDrive uses the format described on Microsoft's [docs](https://docs.microsoft.com/en-us/graph/api/driveitem-delta?view=graph-rest-1.0&tabs=http) site. When those notifications are received in the `onedrive-change-listener`, they get converted to a portable format before sent to the next stage, e.g.:
+Notifications received by a source repository, such as GitHub, OneDrive, Google Drive, etc. have a provider specific format, e.g. OneDrive uses the format described on Microsoft's [docs](https://docs.microsoft.com/en-us/graph/api/driveitem-delta?view=graph-rest-1.0&tabs=http) site. When those notifications are received in the respective listener (e.g. `onedrive-change-listener` for OneDrive), they get converted to a portable format before sent to the next stage, e.g.:
 
 ```
 {
   "type": "onedrive",
-  "provider" {
```
Comment (tripodsan) [inline docs/observation/payload.md]: why did you remove this?

File: docs/observation/payload.md [hunk 2]
```diff
@@ -1,34 +1,52 @@
...
 }
 ```
-The `provider` section contains provider specific information that might be required for downstream handlers.
+The `changes` array contains an entry per change seen since the last batch of changes was sent, e.g. for OneDrive this might look as follows:
+
+```
+{
+  "path": "/myshare/doc1",
+  "time": "2020-03-03T20:05:23Z",
+  "type": "modified",
```
Comment (tripodsan) [inline docs/observation/payload.md]: maybe adding a section that explains the changes in more detail (eg table of properties) and also enumerates the possible values.

File: docs/observation/payload.md [hunk 3]
```diff
@@ -1,34 +1,52 @@
...
+The `provider` section contains provider specific information that might be required for downstream handlers. In the
+example above, a `sourceHash` is included that uniquely identifies items. The `path` property must be available for
+added or modified items, but may be missing for deleted items, if that information is no longer available.
 
 The `mountpoint` contains the information how paths in the changes array should be translated by replacing a prefix matching `root` with `path`, e.g. for the following change:
 ```
 {
-  "id": "...",
-  "path": "/myshare/doc.md",
+  "path": "/myshare/doc",
```
Comment (tripodsan) [inline docs/observation/payload.md]: I think you convinced my to keep the extension.

File: docs/observation/payload.md (hunk 1 above)
Comment (dominique-pfister) [reply docs/observation/payload.md]: It contained the `driveId` only. Since the `sourceHash` (or `uid`) is now part of the payload, the `driveId` is no longer required for calculation, do you still find it necessary to be included?

Comment (tripodsan) [reply docs/observation/payload.md]: well, maybe another consumer of the events might be interested in the drive id.

File: docs/observation/payload.md [hunk 4]
```diff
@@ -1,34 +1,83 @@
...
 
-Notifications received by a source repository, such as GitHub, OneDrive, Google Drive, etc. have a provider specific format, e.g. OneDrive uses the format described on Microsoft's [docs](https://docs.microsoft.com/en-us/graph/api/driveitem-delta?view=graph-rest-1.0&tabs=http) site. When those notifications are received in the `onedrive-change-listener`, they get converted to a portable format before sent to the next stage, e.g.:
+Notifications received by a source repository, such as GitHub, OneDrive, Google Drive, etc. have a provider specific format, e.g. OneDrive uses the format described on Microsoft's [docs](https://docs.microsoft.com/en-us/graph/api/driveitem-delta?view=graph-rest-1.0&tabs=http) site. When those notifications are received in the respective listener (e.g. `onedrive-change-listener` for OneDrive), they get converted to a portable format before sent to the next stage. The format has the following properties:
+
+| Name | Type | Description |
+|------|------|-------------|
+| `type` | _string_ | provider type, possible values are `onedrive`, `github` or `google`. |
+| `owner` | _string_ | GitHub owner |
+| `repo` | _string_ | GitHub repository |
```
Comment (tripodsan) [inline docs/observation/payload.md]:
```suggestion
| `repo` | _string_ | GitHub repository name |
```

File: docs/observation/payload.md [hunk 5]
```diff
@@ -1,34 +1,83 @@
...
+| `ref` | _string_ | GitHub reference |
```
Comment (tripodsan) [inline docs/observation/payload.md]:
```suggestion
| `ref` | _string_ | GitHub reference or branch name |
```

File: docs/observation/payload.md [hunk 6]
```diff
@@ -1,34 +1,83 @@
...
+
+## Changes
+
+The `changes` array contains an entry per change seen since the last batch of changes was sent. They have the following properties:
+
+| Name | Type | Description |
+|------|------|-------------|
+| `path` | _string_ | item path, **required** for _added_ items, **optional** for _modified_ or _deleted_ items |
+| `time` | _string_ | time of change |
+| `type` | _string_ | `modified` for added or modified items, or `deleted` |
```
Comment (tripodsan) [inline docs/observation/payload.md]:
```suggestion
| `type` | _string_ | change type. one of `modified`, 'added', 'deleted'<sup>1<sup> |

<sup>1<sup> If the provider is not able to distinguish between `added` and `modified`, `modified` should be used.
```

File: docs/observation/payload.md [hunk 7]
```diff
@@ -1,34 +1,89 @@
...
-    "root": "/myshare"
-  },
   "changes": [
     ...
   ],
+  "mountpoint": {
+    ...
+  },
+  "provider": {
+    "driveId": "b!-RIj2DuyvEyV1T4NlOaMHk8XkS_I8MdFlUCq1BlcjgmhRfAj3-Z8RY2VpuvV_tpd"
```
Comment (dominique-pfister) [inline docs/observation/payload.md]: Note: that `driveId` is the sample account's root drive available in Microsoft's Graph Explorer
